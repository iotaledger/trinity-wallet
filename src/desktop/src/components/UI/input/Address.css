.address-input {
    display: flex;
    justify-content: space-between;
    width: 100%;
    max-width: 512px;
    margin: 0 auto 30px;
    > div {
        display: block;
        position: relative;
        width: calc(100% - 70px);
    }
    input[type='text'] {
        font-family: 'Lato';
        font-weight: 200;
        font-size: var(--font-size-default);
        line-height: 26px;
        width: 100%;
        border: none;
        background: none;
        color: #fff;
        padding: 4px 0;
        font-size: 18px;
        resize: none;
        &:focus {
            outline: none;
        }
    }
    button {
        padding: 0 8px;
        height: 32px;
        min-width: 0px;
        font-size: 11px;
        margin: 1px 0 0;
        line-height: 18px;
        img {
            width: 14px;
            vertical-align: middle;
            margin: -2px 4px 0 0;
        }
    }
    small {
        position: absolute;
        bottom: 1px;
        left: 0px;
        display: block;
        width: 100%;
        text-align: left;
        &:before,
        &:after {
            position: absolute;
            content: '';
            width: 100%;
            height: 1px;
            background: white;
            top: 0px;
            left: 0px;
        }
        &:after {
            max-width: 0px;
            background: var(--color-yellow);
            transition: all 0.2s ease-out;
        }
    }
    input[type='text']:focus + small:after {
        max-width: 100%;
    }
}

.qrScanner {
    width: 640px;
    max-width: 100%;
    margin: auto;
    text-align: center;
    video {
        transform: rotateY(180deg);
        max-width: 100%;
        margin-bottom: 50px;
        border: 2px solid #fff;
    }
}
