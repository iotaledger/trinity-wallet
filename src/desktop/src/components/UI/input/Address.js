import React from 'react';
import QrReader from 'react-qr-reader';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { ADDRESS_LENGTH } from 'libs/util';
import css from './Input.css';
import Modal from 'components/UI/Modal';

import Camera from 'images/camera.png';

export default class AddressInput extends React.PureComponent {
    static propTypes = {
        address: PropTypes.string.isRequired,
        label: PropTypes.string.isRequired,
        closeLabel: PropTypes.string.isRequired,
    };

    state = {
        showScanner: false,
    };

    openScanner = e => {
        e.preventDefault();
        this.setState(() => ({
            showScanner: true,
        }));
    };

    closeScanner = e => {
        e.preventDefault();
        this.setState(() => ({
            showScanner: false,
        }));
    };

    onScanEvent = address => {
        if (address !== null) {
            this.setState(() => ({
                showScanner: false,
            }));
            this.props.onChange(address);
        }
    };

    onScanError = err => {
        console.log(err);
    };

    render() {
        const { address, label, closeLabel } = this.props;
        const { showScanner } = this.state;

        return (
            <div className={css.input}>
                <fieldset>
                    <a onClick={this.openScanner}>
                        <img src={Camera} alt="" />
                    </a>
                    <input
                        type="text"
                        value={address}
                        onChange={e => this.props.onChange(e.target.value)}
                        maxLength={ADDRESS_LENGTH}
                    />
                    <small>{label}</small>
                </fieldset>
                {showScanner && (
                    <Modal isOpen={true} onStateChange={showScanner => this.setState({ showScanner })} hideCloseButton>
                        <div className={css.qrScanner}>
                            <QrReader delay={350} onError={this.onScanError} onScan={this.onScanEvent} />
                            <Button type="button" onClick={this.closeScanner} variant="cta">
                                {closeLabel}
                            </Button>
                        </div>
                    </Modal>
                )}
            </div>
        );
    }
}
