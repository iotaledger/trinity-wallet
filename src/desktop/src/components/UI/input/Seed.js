import React from 'react';
import QrReader from 'react-qr-reader';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { MAX_SEED_LENGTH } from 'libs/util';
import css from './Input.css';
import Modal from 'components/UI/Modal';

import Camera from 'images/camera.png';

export default class SeedInout extends React.PureComponent {
    static propTypes = {
        seed: PropTypes.string.isRequired,
        label: PropTypes.string.isRequired,
        closeLabel: PropTypes.string.isRequired,
    };

    state = {
        showScanner: false,
    };

    openScanner = e => {
        e.preventDefault();
        this.setState(() => ({
            showScanner: true,
        }));
    };

    closeScanner = e => {
        e.preventDefault();
        this.setState(() => ({
            showScanner: false,
        }));
    };

    onScanEvent = address => {
        if (address !== null) {
            this.setState(() => ({
                showScanner: false,
            }));
            this.props.onChange(address);
        }
    };

    onScanError = err => {
        console.log(err);
    };

    render() {
        const { seed, label, closeLabel } = this.props;
        const { showScanner } = this.state;

        return (
            <div className={css.input}>
                <fieldset>
                    <a onClick={this.openScanner}>
                        <img src={Camera} alt="" />
                    </a>
                    <input
                        type="text"
                        value={seed}
                        onChange={e => this.props.onChange(e.target.value)}
                        maxLength={MAX_SEED_LENGTH}
                    />
                    <small>{label}</small>
                </fieldset>
                {showScanner && (
                    <Modal isOpen={true} onStateChange={showScanner => this.setState({ showScanner })} hideCloseButton>
                        <div className={css.qrScanner}>
                            <QrReader delay={350} onError={this.onScanError} onScan={this.onScanEvent} />
                            <Button type="button" onClick={this.closeScanner} variant="secondary">
                                {closeLabel}
                            </Button>
                        </div>
                    </Modal>
                )}
            </div>
        );
    }
}
