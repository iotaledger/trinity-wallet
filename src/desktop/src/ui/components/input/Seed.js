import React from 'react';
import QrReader from 'react-qr-reader';
import PropTypes from 'prop-types';
import { getChecksum, MAX_SEED_LENGTH } from 'libs/iota/utils';

import Modal from 'ui/components/modal/Modal';
import Button from 'ui/components/Button';
import Icon from 'ui/components/Icon';
import css from './input.css';

/**
 * Seed input component
 */
export default class SeedInput extends React.PureComponent {
    static propTypes = {
        /** Current seed value */
        seed: PropTypes.string.isRequired,
        /** Seed input label */
        label: PropTypes.string.isRequired,
        /** Camera modal close label */
        closeLabel: PropTypes.string.isRequired,
        /** Seed change event function
         * @param {string} value - Current seed value
         */
        onChange: PropTypes.func.isRequired,
    };

    state = {
        showScanner: false,
    };

    onScanEvent = (address) => {
        if (address !== null) {
            this.setState(() => ({
                showScanner: false,
            }));
            this.props.onChange(address);
        }
    };

    closeScanner = () => {
        this.setState(() => ({
            showScanner: false,
        }));
    };

    openScanner = () => {
        this.setState(() => ({
            showScanner: true,
        }));
    };

    render() {
        const { seed, label, closeLabel, onChange } = this.props;
        const { showScanner } = this.state;

        return (
            <div className={css.input}>
                <fieldset>
                    <a onClick={this.openScanner}>
                        <Icon icon="camera" size={16} />
                    </a>
                    <input
                        type="text"
                        value={seed}
                        onChange={(e) => onChange(e.target.value.toUpperCase())}
                        maxLength={MAX_SEED_LENGTH}
                    />
                    <small>{label}</small>
                </fieldset>
                {seed.length ? (
                    <span className={css.info}>{seed.length < MAX_SEED_LENGTH ? '< 81' : getChecksum(seed)}</span>
                ) : null}
                {showScanner && (
                    <Modal isOpen onClose={this.closeScanner}>
                        <div className={css.qrScanner}>
                            <QrReader delay={350} onScan={this.onScanEvent} />
                            <Button type="button" onClick={this.closeScanner} variant="secondary">
                                {closeLabel}
                            </Button>
                        </div>
                    </Modal>
                )}
            </div>
        );
    }
}
