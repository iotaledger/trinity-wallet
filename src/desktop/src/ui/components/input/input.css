.input {
    display: block;
    width: 100%;
    padding: 25px 0 30px;
    text-align: right;
    fieldset {
        padding: 0px;
        margin: 0px;
        border: none;
        position: relative;
        text-align: left;
        input,
        textarea,
        select {
            font-family: 'Lato';
            font-weight: 200;
            font-size: var(--font-size-default);
            line-height: 26px;
            width: 100%;
            border: 1px solid var(--input-border);
            background: var(--input-bg);
            color: var(--input);
            padding: 6px 12px;
            resize: none;
            -webkit-appearance: none;
            border-radius: 4px;
            &:focus {
                outline: none;
            }
            &:disabled {
                pointer-events: none;
                opacity: 0.5;
            }
        }
        select {
            position: relative;
            &:after {
                content: '';
                display: block;
                width: 0;
                height: 0;
                border-style: solid;
                border-width: 6px 4px 0 4px;
                border-color: var(--body) transparent transparent transparent;
                position: absolute;
                right: 12px;
                top: calc(50% - 2px);
                opacity: 0.6;
                pointer-event: none;
            }
        }
        small {
            position: absolute;
            white-space: nowrap;
            left: 0px;
            top: -20px;
            font-size: var(--font-size-xs);
            color: var(--label);
            text-transform: uppercase;
            transition: color 0.2s ease-out;
        }
        input:focus + small,
        textarea:focus + small,
        select:focus + small {
            color: var(--label-hover);
        }
        a {
            display: flex;
            position: absolute;
            align-items: center;
            justify-content: center;
            position: absolute;
            height: 100%;
            min-width: 35px;
            top: 0px;
            right: 0px;
            cursor: pointer;
            user-select: none;
            color: var(--input-alt);
            &:before {
                content: '';
                position: absolute;
                width: 2px;
                height: 22px;
                top: 50%;
                left: -2px;
                border-radius: 4px;
                background: var(--input-alt);
                transform: translate(0, -50%);
            }
            > strong {
                font-size: 11px;
                font-weight: 200;
                > span {
                    margin-left: 4px;
                }
            }
            &:hover {
                color: var(--input);
            }
        }
        a + input {
            padding-right: 40px;
        }
    }
}

.info {
    display: block;
    float: right;
    font-size: 11px;
    color: var(--input);
    line-height: 14px;
    padding: 4px 8px 2px;
    border-radius: 0 0 4px 4px;
    position: relative;
    background: var(--input-bg);
    margin: -1px 38px 0 0;
    border: 1px solid var(--input-border);
    border-top: none;
}

.checkbox {
    display: block;
    float: right;
    margin: 12px 3px 0 auto;
    font-size: 10px;
    line-height: 14px;
    padding: 0 38px 2px 0;
    position: relative;
    text-transform: uppercase;
    cursor: pointer;
    &:after {
        content: '';
        width: 28px;
        height: 15px;
        position: absolute;
        top: 0px;
        right: 1px;
        border-radius: 15px;
        border: 1px solid var(--body);
    }
    &:before {
        content: '';
        width: 9px;
        height: 9px;
        position: absolute;
        top: 3px;
        right: 16px;
        border-radius: 9px;
        border: 1px solid var(--body);
        transition: all 0.2s cubic-bezier(0, 0, 0.2, 1);
    }
    &.on {
        color: var(--primary);
        &:before {
            border-color: var(--primary);
            background: var(--primary);
            right: 4px;
        }
    }
}

.strike {
    &:after {
        content: '';
        position: absolute;
        width: 1px;
        height: 16px;
        border-radius: 2px;
        background: var(--input-alt);
        transform: rotate(45deg);
    }
    &:hover:after {
        background: var(--input);
    }
}

.down {
    position: relative;
    opacity: 1;
    z-index: 2;
    pointer-events: none;
    &:before {
        display: none;
    }
    &:after {
        content: '';
        display: block;
        width: 0;
        height: 0;
        border-style: solid;
        border-width: 6px 4px 0 4px;
        border-color: var(--input) transparent transparent transparent;
        position: absolute;
        right: 12px;
        top: calc(50% - 2px);
    }
}

.qrScanner {
    display: block;
    margin: 0 auto;
    width: 320px;
    text-align: center;
    video {
        transform: rotateY(180deg);
        max-width: 100%;
        margin-bottom: 50px;
        border: 2px solid var(--input);
    }
    > section {
        margin-bottom: 40px;
    }
}
