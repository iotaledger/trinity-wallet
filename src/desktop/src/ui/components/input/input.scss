.input {
    display: block;
    width: 100%;
    max-width: 520px;
    margin: 0 auto;
    padding: 25px 0 30px;
    transition: opacity 0.4s cubic-bezier(0.165, 0.84, 0.44, 1);
    position: relative;

    fieldset {
        border: none;
        position: relative;
        input,
        textarea,
        select,
        .editable {
            font-family: 'SourceSansPro';
            font-weight: 300;
            font-size: var(--font-size-default);
            line-height: 46px;
            width: 100%;
            border: 1px solid var(--input-border);
            background: var(--input-bg);
            color: var(--input);
            padding: 0 20px 0 14px;
            resize: none;
            -webkit-appearance: none;
            border-radius: 4px;
            transition: border 0.4s cubic-bezier(0.165, 0.84, 0.44, 1);
            &:focus {
                outline: none;
                border-color: var(--input-focus);
            }
            &:disabled {
                pointer-events: none;
                opacity: 0.5;
            }
        }
        .editable {
            display: block;
            font-family: 'SourceCodePro';
            white-space: nowrap;
            overflow: hidden;
            height: 46px;
            text-align: left;
            max-width: 520px;
            padding: 0 40px 0 0;
            > div {
                padding: 0 20px;
                overflow: hidden;

                span {
                    display: inline-block;
                    position: relative;
                    height: 100%;
                    &:before {
                        display: inline-block;
                        position: absolute;
                        left: 0px;
                        content: attr(data-letter);
                        pointer-events: none;
                        line-height: 44px;
                    }
                }
            }
        }
        select {
            position: relative;
            &:after {
                content: '';
                display: block;
                width: 0;
                height: 0;
                border-style: solid;
                border-width: 6px 4px 0 4px;
                border-color: var(--input-alt) transparent transparent transparent;
                position: absolute;
                right: 12px;
                top: calc(50% - 2px);
                pointer-event: none;
            }
        }
        small {
            position: absolute;
            white-space: nowrap;
            left: 0px;
            top: -24px;
            color: var(--label);
            text-transform: uppercase;
            transition: color 0.2s ease-out;
            font-size: 13px;
            letter-spacing: 0.2px;
        }
        input:focus + small,
        textarea:focus + small,
        select:focus + small {
            color: var(--label-hover);
        }
        a {
            display: flex;
            position: absolute;
            align-items: center;
            justify-content: center;
            position: absolute;
            height: 100%;
            min-width: 35px;
            top: 0px;
            right: 0px;
            cursor: pointer;
            user-select: none;
            color: var(--input-alt);
            &:before {
                content: '';
                position: absolute;
                width: 2px;
                height: 22px;
                top: 50%;
                left: -2px;
                border-radius: 4px;
                background: var(--input-alt);
                transform: translate(0, -50%);
            }
            > strong {
                font-size: 11px;
                > span {
                    margin-left: 4px;
                }
            }
            &:hover {
                color: var(--input);
            }
            &.right {
                left: 0px;
                right: auto;
                background: red;
                z-index: 2;
                width: 42px;
                background: var(--input-bg);
                border-radius: 4px 0 0 4px;
                border: 1px solid var(--input-border);
                border-right: none;
                &:before {
                    display: none;
                }
            }
        }
        p {
            display: block;
            position: absolute;
            top: 0px;
            right: 0px;
            right: 50px;
            line-height: 38px;
            white-space: nowrap;
            color: var(--input-alt);
            pointer-events: none;
        }
        a + input {
            padding-right: 40px;
        }
    }

    h5 {
        display: block;
        width: 100%;
        border: 1px dashed var(--body-alt);
        font-size: 12px;
        line-height: 12px;
        margin: 32px 0 0 0;
        padding: 24px 0;
        font-size: 14px;
        text-align: center;
        cursor: pointer;

        svg {
            display: block;
            margin: 0 auto 20px;
            pointer-events: none;
        }
        span {
            display: block;
            pointer-events: none;
            line-height: 18px;
        }
    }

    > fieldset > strong {
        display: none;
        position: absolute;
        top: 58px;
        left: 8px;
        text-align: left;
        pointer-events: none;
        font-weight: 300;
        font-size: 12px;
        line-height: 12px;
        color: var(--body);
        span {
            float: left;
            margin: -1px 4px 0 0;
        }
    }

    &.disabled {
        opacity: 0.25;
        pointer-events: none;
    }

    &.padded {
        margin-bottom: 18px;
    }

    &.seed {
        fieldset .editable {
            padding: 0 32px 0 21px;
        }
    }

    &.capsLock {
        input:focus + small + strong {
            display: block;
        }
        > :before {
            content: '';
            display: block;
            position: absolute;
            top: 1px;
            right: 40px;
            width: 46px;
            height: 46px;
            background: var(--input-bg);
        }
    }
}

.info {
    display: block;
    position: absolute;
    top: 71px;
    right: -5px;
    font-size: 11px;
    color: var(--input);
    line-height: 14px;
    padding: 4px 8px 2px;
    border-radius: 0 0 4px 4px;
    background: var(--input-bg);
    margin: -1px 38px 0 0;
    border: 1px solid var(--input-border);
    border-top: none;
}

.checkbox {
    display: block;
    position: absolute;
    bottom: -10px;
    right: 0px;
    float: right;
    font-size: 10px;
    line-height: 14px;
    padding: 0 38px 2px 0;
    text-transform: uppercase;
    cursor: pointer;
    &:after {
        content: '';
        width: 28px;
        height: 15px;
        position: absolute;
        top: 0px;
        right: 1px;
        border-radius: 15px;
        border: 1px solid var(--body);
    }
    &:before {
        content: '';
        width: 9px;
        height: 9px;
        position: absolute;
        top: 3px;
        right: 16px;
        border-radius: 9px;
        border: 1px solid var(--body);
        transition: all 0.2s cubic-bezier(0, 0, 0.2, 1);
    }
    &.on {
        color: var(--primary);
        &:before {
            border-color: var(--primary);
            background: var(--primary);
            right: 4px;
        }
    }
}

.dropdown {
    position: absolute;
    top: 0;
    left: 0px;
    width: 100%;
    padding: 38px 0 0 0;
    z-index: 2;
    max-height: 40px;
    overflow: hidden;
    li {
        display: block;
        list-style: none;
        padding-left: 8px;
        width: 100%;
        text-align: left;
        line-height: 28px;
        background: var(--input-bg);
        color: var(--input-alt);
        &.selected {
            color: var(--input);
        }
        &:last-of-type {
            padding-bottom: 8px;
            border-radius: 0 0 4px 4px;
        }
        &:hover {
            color: var(--input);
        }
    }
    &:hover {
        max-height: none;
    }
}

.strike {
    &:after {
        content: '';
        position: absolute;
        width: 1px;
        height: 16px;
        border-radius: 2px;
        background: var(--input-alt);
        transform: rotate(45deg);
    }
    &:hover:after {
        background: var(--input);
    }
}

.down {
    position: relative;
    z-index: 2;
    pointer-events: none;
    &:before {
        display: none;
    }
    &:after {
        content: '';
        display: block;
        width: 0;
        height: 0;
        border-style: solid;
        border-width: 6px 4px 0 4px;
        border-color: var(--input) transparent transparent transparent;
        position: absolute;
        right: 12px;
        top: calc(50% - 2px);
    }
}

.qrScanner {
    display: block;
    margin: 0 auto;
    width: 320px;
    text-align: center;
    video {
        transform: rotateY(180deg);
        max-width: 100%;
        margin-bottom: 50px;
        border: 2px solid var(--input);
    }
    > section {
        margin-bottom: 40px;
    }
}

.score {
    display: flex;
    position: absolute;
    top: -17px;
    right: 0px;
    span {
        display: block;
        height: 4px;
        width: 50px;
        background: var(--body-alt);
        margin-left: 2px;
    }
    &[data-strength='1'] {
        span:nth-child(1) {
            background: var(--negative);
        }
    }
    &[data-strength='3'],
    &[data-strength='2'],
    &[data-strength='1'] {
        span:nth-child(1),
        span:nth-child(2) {
            background: var(--negative);
        }
    }
    &[data-strength='4'] {
        span {
            background: var(--positive);
        }
    }
}

.hint {
    position: absolute;
    bottom: -28px;
    right: 0;
    font-size: 14px;
}

.valid {
    position: absolute;
    top: 10px;
    right: -40px;
    opacity: 0.25;
    &.isValid {
        opacity: 1;
    }
}
