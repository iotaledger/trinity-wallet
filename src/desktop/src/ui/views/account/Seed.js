/* global Electron */
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { translate } from 'react-i18next';
import { connect } from 'react-redux';

import { MAX_SEED_LENGTH } from 'libs/iota/utils';
import { byteToChar } from 'libs/crypto';

import { getSelectedAccountName } from 'selectors/accounts';

import Button from 'ui/components/Button';
import Clipboard from 'ui/components/Clipboard';
import ModalPassword from 'ui/components/modal/Password';
import SeedPrint from 'ui/components/SeedPrint';

import css from './seed.scss';

/**
 * Account Seed settings component
 */
class Seed extends PureComponent {
    static propTypes = {
        /** Current account name */
        accountName: PropTypes.string.isRequired,
        /** Translation helper
         * @param {String} translationString - Locale string identifier to be translated
         * @ignore
         */
        t: PropTypes.func.isRequired,
    };

    state = {
        hidden: true,
        seed: null,
    };

    render() {
        const { accountName, t } = this.props;
        const { seed, hidden } = this.state;

        if (!hidden && !seed) {
            return (
                <ModalPassword
                    isOpen
                    inline
                    onSuccess={(password, seed) => this.setState({ seed })}
                    onClose={() => this.setState({ hidden: true })}
                    seedName={accountName}
                    content={{
                        title: t('viewSeed:enterPassword'),
                        confirm: t('accountManagement:viewSeed'),
                    }}
                />
            );
        }

        const checksum = seed ? Electron.getChecksum(seed) : '';

        return (
            <form>
                <p className={css.seed}>
                    <Clipboard
                        text={seed || ''}
                        title={t('copyToClipboard:seedCopied')}
                        success={t('copyToClipboard:seedCopiedExplanation')}
                    >
                        {seed && !hidden
                            ? seed.map((byte, index) => {
                                  if (index % 3 !== 0) {
                                      return null;
                                  }
                                  const letter = byteToChar(byte % 27);
                                  return (
                                      <React.Fragment key={`${index}${letter}`}>
                                          {letter}
                                          {byteToChar(seed[index + 1])}
                                          {byteToChar(seed[index + 2])}{' '}
                                      </React.Fragment>
                                  );
                              })
                            : new Array(MAX_SEED_LENGTH / 3).join('... ')}
                    </Clipboard>
                </p>
                <fieldset>
                    <Button onClick={() => this.setState({ hidden: !hidden })}>
                        {hidden ? t('settings:show') : t('settings:hide')}
                    </Button>
                    {!hidden && <Button onClick={() => window.print()}>{t('paperWallet')}</Button>}
                </fieldset>
                {!hidden && <SeedPrint seed={seed} checksum={checksum} filled />}
            </form>
        );
    }
}

const mapStateToProps = (state) => ({
    accountName: getSelectedAccountName(state),
});

export default connect(mapStateToProps)(translate()(Seed));
