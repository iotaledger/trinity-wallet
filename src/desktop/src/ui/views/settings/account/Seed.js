/* global Electron */
import React, { PureComponent, Fragment } from 'react';
import PropTypes from 'prop-types';
import { withTranslation } from 'react-i18next';

import { MAX_SEED_LENGTH } from 'libs/iota/utils';
import { byteToChar, capitalize } from 'libs/iota/converter';
import SeedStore from 'libs/SeedStore';

import Button from 'ui/components/Button';
import Modal from 'ui/components/modal/Modal';
import ModalPassword from 'ui/components/modal/Password';
import SeedPrint from 'ui/components/SeedPrint';
import SeedExport from 'ui/global/SeedExport';

import css from './seed.scss';
import cssIndex from '../index.scss';

/**
 * Account Seed settings component
 */
class Seed extends PureComponent {
    static propTypes = {
        /** @ignore */
        account: PropTypes.object.isRequired,
        /** @ignore */
        t: PropTypes.func.isRequired,
    };

    state = {
        action: null,
        seed: false,
    };

    componentDidMount() {
        this.seed = [];
    }

    /**
     * Trigger seed print after component is updated
     */
    componentDidUpdate(_prevProps, prevState) {
        if (this.state.action === 'print' && !prevState.seed && this.state.seed) {
            window.print();
        }
    }

    componentWillUnmount() {
        for (let i = 0; i < this.seed.length * 3; i++) {
            this.seed[i % this.seed.length] = 0;
        }

        Electron.garbageCollect();
    }

    /**
     * Retrieve seed and set to state
     */
    setSeed = async (password) => {
        const { accountName, meta } = this.props.account;

        const seedStore = await new SeedStore[meta.type](password, accountName, meta);
        this.seed = await seedStore.getSeed();

        this.setState({
            seed: true,
        });
    };

    render() {
        const { t } = this.props;
        const { meta, accountName } = this.props.account;
        const { seed, action } = this.state;

        if (meta && !SeedStore[meta.type].isSeedAvailable) {
            return (
                <div className={cssIndex.scroll}>
                    <article>
                        <h3>{t('viewSeed:notAvailable', { accountType: capitalize(meta.type) })}</h3>
                        <p>{t(`viewSeed:notAvailableExplanation${capitalize(meta.type)}`)}</p>
                        {typeof meta.index === 'number' && (
                            <Fragment>
                                <hr />
                                <p>
                                    {t('viewSeed:accountIndex')}: <strong>{meta.index}</strong>
                                </p>
                            </Fragment>
                        )}
                        {typeof meta.page === 'number' && meta.page > 0 && (
                            <Fragment>
                                <hr />
                                <p>
                                    <hr />
                                    {t('viewSeed:accountPage')}: <strong>{meta.page}</strong>
                                </p>
                            </Fragment>
                        )}
                    </article>
                </div>
            );
        }

        if (action && !seed) {
            return (
                <ModalPassword
                    isOpen
                    inline
                    onSuccess={this.setSeed}
                    onClose={() => this.setState({ action: null })}
                    seedName={accountName}
                    content={{
                        title: action === 'view' ? t('viewSeed:enterPassword') : t('login:enterPassword'),
                        confirm:
                            action === 'view'
                                ? t('accountManagement:viewSeed')
                                : action === 'export'
                                ? t('seedVault:exportSeedVault')
                                : t('paperWallet'),
                    }}
                />
            );
        }

        const checksum = seed ? Electron.getChecksum(this.seed) : '';

        return (
            <React.Fragment>
                <form>
                    <fieldset>
                        <p className={css.seed}>
                            <span>
                                {seed && action === 'view'
                                    ? this.seed.map((byte, index) => {
                                          if (index % 3 !== 0) {
                                              return null;
                                          }
                                          const letter = byteToChar(byte);
                                          return (
                                              <React.Fragment key={`${index}${letter}`}>
                                                  {letter}
                                                  {byteToChar(this.seed[index + 1])}
                                                  {byteToChar(this.seed[index + 2])}{' '}
                                              </React.Fragment>
                                          );
                                      })
                                    : new Array(MAX_SEED_LENGTH / 3).join('... ')}
                            </span>
                            {seed && action === 'view' && (
                                <small>
                                    {t('checksum')}: <strong>{checksum}</strong>
                                </small>
                            )}
                        </p>
                    </fieldset>
                    <fieldset>
                        <Button
                            className="small"
                            onClick={() => this.setState({ action: action !== 'view' ? 'view' : null })}
                        >
                            {action === 'view' ? t('settings:hide') : t('settings:show')}
                        </Button>
                        <Button
                            className="small"
                            onClick={() => (!seed ? this.setState({ action: 'print' }) : window.print())}
                        >
                            {t('paperWallet')}
                        </Button>
                        <Button className="small" onClick={() => this.setState({ action: 'export' })}>
                            {t('seedVault:exportSeedVault')}
                        </Button>
                    </fieldset>
                    <SeedPrint seed={this.seed} checksum={checksum} filled />
                </form>
                <Modal
                    variant="fullscreen"
                    isOpen={seed && action === 'export'}
                    onClose={() => this.setState({ action: null })}
                >
                    <SeedExport
                        seed={this.seed || []}
                        title={accountName}
                        onClose={() => this.setState({ action: null })}
                    />
                </Modal>
            </React.Fragment>
        );
    }
}

export default withTranslation()(Seed);
