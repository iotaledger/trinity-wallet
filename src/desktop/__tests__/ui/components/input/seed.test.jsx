import React from 'react';
import { shallow } from 'enzyme';

import { SeedComponent as Seed } from 'ui/components/input/Seed';

const props = {
    seed: [0, 1, 2],
    label: 'Foo',
    closeLabel: 'Bar',
    onChange: jest.fn(),
    setAccountInfoDuringSetup: () => {},
    updateImportName: false,
    generateAlert: () => {},
    t: (str) => str,
};

global.Electron = {
    getChecksum: () => 'XXX',
};

describe('Seed input component', () => {
    test('Render the component', () => {
        const wrapper = shallow(<Seed {...props} />);

        expect(wrapper).toMatchSnapshot();
    });

    test('Input value', () => {
        const wrapper = shallow(<Seed {...props} />);

        expect(wrapper.find('.letter')).toHaveLength(3);
    });

    test('Input change callback', () => {
        const wrapper = shallow(<Seed {...props} />);

        wrapper.find('[contentEditable]').simulate('keyDown', { preventDefault: () => {}, key: 'a', keyCode: 97 });
        expect(props.onChange).toHaveBeenLastCalledWith([0, 1, 2, 1]);
    });

    test('Checksum display', () => {
        const testValues = [Array(20).fill(1), Array(81).fill(1), Array(90).fill(1)];
        const testChecksums = ['< 81', 'XXX', '> 81'];

        testValues.forEach((testValue, index) => {
            const wrapper = shallow(<Seed {...props} seed={testValue} />);
            expect(wrapper.find('.info').text()).toEqual(testChecksums[index]);
        });
    });

    test('Input label', () => {
        const wrapper = shallow(<Seed {...props} />);

        expect(wrapper.find('small').text()).toEqual('Foo');
    });
});
