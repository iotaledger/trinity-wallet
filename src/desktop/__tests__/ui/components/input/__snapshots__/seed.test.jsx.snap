// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Seed input component Render the component 1`] = `
<div
  className="input seed"
>
  <fieldset>
    <a
      className="strike"
      onClick={[Function]}
    >
      <Icon
        icon="eye"
        size={16}
      />
    </a>
    <a
      className="right"
      onClick={[Function]}
    >
      <Icon
        icon="camera"
        size={16}
      />
    </a>
    <div
      className="editable"
    >
      <div
        contentEditable={true}
        onKeyDown={[Function]}
        onPaste={[Function]}
        suppressContentEditableWarning={true}
      >
        <span
          className="letter"
          data-letter="•"
          key="0-0"
        >
           
        </span>
        <span
          className="letter"
          data-letter="•"
          key="1-1"
        >
           
        </span>
        <span
          className="letter"
          data-letter="•"
          key="2-2"
        >
           
        </span>
      </div>
    </div>
    <small>
      Foo
    </small>
  </fieldset>
  <Connect(withI18nextTranslation(Dropzone))
    onDrop={[Function]}
    onTextDrop={[Function]}
  />
  <span
    className="info"
  >
    &lt; 81
  </span>
</div>
`;
